pipeline/id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/steps/SourceIdentifierScanner.scala

```diff
@@ -0,0 +1,108 @@
+
+/**
+  * SourceIdentifierScanner provides 2 methods:
+  *
+  * - `scan` takes Json and returns all of the sourceIdentifiers that are in it
+  * - `update` takes Json and a map of (SourceIdentifier -> Identifier) and adds
+  *   a canonicalId field next to sourceIdentifiers, as well as replacing
+  *   `identifiedType` fields with `type` fields of the same value.
+  *
+  */
+object SourceIdentifierScanner extends Logging {
+  def update(inputJson: Json,
+             identifiers: Map[SourceIdentifier, Identifier]): Try[Json] =
+    Try {
+      val updateNode =
+        (updateNodeType _) compose addCanonicalIdToNode(identifiers)
+      // Plated transforms operate on self-similar *children*
+      // so we need to update the root separately
+      val updatedRoot = updateNode(inputJson)
+      Plated.transform[Json](updateNode)(updatedRoot)
+    }
+
+  private def updateNodeType(node: Json): Json =
+    root.identifiedType.json
+      .getOption(node)
+      .flatMap(_.asString)
+      .map { identifiedType =>
+        root.obj.modify { obj =>
+          ("type", Json.fromString(identifiedType)) +:
+            obj.remove("identifiedType")
+        }(node)
+      }
+      .getOrElse(node)
+
+  private def addCanonicalIdToNode(
+    identifiers: Map[SourceIdentifier, Identifier])(node: Json): Json =
+    root.sourceIdentifier.json
+      .getOption(node)
+      .map(parseSourceIdentifier)
+      .map(getCanonicalId(identifiers))
+      .map { canonicalId =>
+        root.obj.modify { obj =>
+          ("canonicalId", Json.fromString(canonicalId)) +: obj
+        }(node)
+      }
+      .getOrElse(node)
+
+  private def getCanonicalId(identifiers: Map[SourceIdentifier, Identifier])(
+    sourceIdentifier: SourceIdentifier): String =
+    identifiers
+      .getOrElse(
+        sourceIdentifier,
+        throw new RuntimeException(
+          s"Could not find $sourceIdentifier in $identifiers")
+      )
+      .CanonicalId
+
+  def scan(inputJson: Json): Try[List[SourceIdentifier]] =
+    Try(
+      iterate(
+        root.each.json.getAll(inputJson),
+        root.sourceIdentifier.json
+          .getOption(inputJson)
+          .map(parseSourceIdentifier)
+          .toList
+      )
+    )
+
+  @tailrec
+  private def iterate(
+    children: List[Json],
+    identifiers: List[SourceIdentifier]): List[SourceIdentifier] =
+    children match {
+      case Nil => identifiers
+      case headChild :: tailChildren =>
+        iterate(
+          root.each.json.getAll(headChild) ++ tailChildren,
+          identifiers ++ root.sourceIdentifier.json
+            .getOption(headChild)
+            .map(parseSourceIdentifier)
+        )
+    }
+
+  private def parseSourceIdentifier(
+    sourceIdentifierJson: Json): SourceIdentifier = {
+    sourceIdentifierJson.as[SourceIdentifier].toTry match {
+      case Success(sourceIdentifier) => sourceIdentifier
+      case Failure(exception) =>
+        error(s"Error parsing source identifier: $exception")
```

Comment:
Is it worth logging the value of `sourceIdentifierJson`? My experience of Circe error messages is that they’re only helpful if you can see the JSON it was looking at.